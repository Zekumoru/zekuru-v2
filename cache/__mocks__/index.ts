import { sampleChannelLinks, sampleTranslateChannels } from './sample-data';

export = {
  translateChannel: {
    get: jest
      .fn()
      .mockImplementation(async (channelId) =>
        sampleTranslateChannels.find((trChannel) => trChannel.id === channelId)
      ),
    set: jest.fn(),
  },
  channelLink: {
    create: jest
      .fn()
      .mockImplementation(async (channelId) =>
        sampleChannelLinks.find((chLink) => chLink.id === channelId)
      ),
    get: jest
      .fn()
      .mockImplementation(async (channelId) =>
        sampleChannelLinks.find((chLink) => chLink.id === channelId)
      ),
    update: jest.fn(),
  },
};
