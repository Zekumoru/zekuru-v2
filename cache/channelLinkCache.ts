import { Collection } from 'discord.js';
import ChannelLink, { IChannelLink } from '../db/models/ChannelLink';

const cacheLinks = new Collection<string, IChannelLink>();

const fetchOrCreateLink = async (channelId: string, guildId: string) => {
  // fetch from db, if not exists, create new
  const link = await ChannelLink.findOne({ id: channelId })
    .populate('links')
    .exec();
  if (link) return link;

  const newLink = new ChannelLink({
    // guildId,
    id: channelId,
    links: [],
  });
  await newLink.save();
  return newLink;
};

const create = async (channelId: string, guildId: string) => {
  // create otherwise fetch
  const link = await fetchOrCreateLink(channelId, guildId);

  cacheLinks.set(channelId, {
    _id: link._id,
    id: link.id,
    guildId: link.guildId,
    links: link.links,
    createdAt: link.createdAt,
  });
  return cacheLinks.get(channelId)!;
};

const get = async (channelId: string) => {
  // get from cache
  const linkCache = cacheLinks.get(channelId);
  if (linkCache) return linkCache;

  // if not exists, fetch from db and set to cache
  const link = await ChannelLink.findOne({ id: channelId })
    .populate('links')
    .exec();
  if (!link) return null;

  cacheLinks.set(channelId, {
    _id: link._id,
    id: link.id,
    guildId: link.guildId,
    links: link.links,
    createdAt: link.createdAt,
  });
  return cacheLinks.get(channelId)!;
};

const update = async (channelLink: IChannelLink) => {
  // update in db
  const link = await fetchOrCreateLink(channelLink.id, channelLink.guildId);
  link.overwrite(channelLink);
  await link.save();

  // update cache
  cacheLinks.set(channelLink.id, {
    _id: link._id,
    id: link.id,
    guildId: link.guildId,
    links: link.links,
    createdAt: link.createdAt,
  });
};

const deleteLink = async (channelId: string) => {
  // check if it exists first
  if ((await get(channelId)) == null) return;

  // delete from db
  await ChannelLink.deleteOne({ id: channelId });

  // remove from cache
  cacheLinks.delete(channelId);
};

const clear = () => {
  cacheLinks.clear();
};

export default { create, update, get, delete: deleteLink, clear };
