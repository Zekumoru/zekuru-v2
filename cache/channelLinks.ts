import { Collection } from 'discord.js';
import ChannelLink, { IChannelLink } from '../models/ChannelLink';

const cacheLinks = new Collection<string, IChannelLink>();

const getLink = async (channelId: string, guildId: string) => {
  // fetch from db, if not exists, create new
  const link = await ChannelLink.findOne({ id: channelId })
    .populate('links')
    .exec();
  if (link) return link;

  const newLink = new ChannelLink({
    guildId,
    id: channelId,
    links: [],
  });
  await newLink.save();
  return newLink;
};

const create = async (channelId: string, guildId: string) => {
  // create otherwise fetch
  const link = await getLink(channelId, guildId);

  cacheLinks.set(channelId, {
    guildId,
    _id: link._id,
    id: link.id,
    links: link.links,
    createdAt: link.createdAt,
  });
  return cacheLinks.get(channelId)!;
};

const update = async (channelLink: IChannelLink) => {
  // update in db
  const link = await getLink(channelLink.id, channelLink.guildId);
  link.overwrite(channelLink);
  await link.save();

  // update cache
  cacheLinks.set(
    channelLink.id,
    await getLink(channelLink.id, channelLink.guildId)
  );
};

const get = async (channelId: string) => {
  // get from cache
  const linkCache = cacheLinks.get(channelId);
  if (linkCache) return linkCache;

  // if not exists, fetch from db and set to cache
  const link = await ChannelLink.findOne({ id: channelId })
    .populate('links')
    .exec();
  if (!link) return null;

  cacheLinks.set(channelId, {
    _id: link._id,
    id: link.id,
    guildId: link.guildId,
    links: link.links,
    createdAt: link.createdAt,
  });
  return cacheLinks.get(channelId)!;
};

const deleteLink = async (channelId: string) => {
  // check if it exists first
  if ((await get(channelId)) == null) return;

  // delete from db
  await ChannelLink.deleteOne({ id: channelId });

  // remove from cache
  cacheLinks.delete(channelId);
};

export default { create, update, get, delete: deleteLink };
