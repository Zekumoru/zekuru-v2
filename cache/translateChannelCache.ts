import { Collection } from 'discord.js';
import type { ITranslateChannel } from '../db/models/TranslateChannel';
import TranslateChannel from '../db/models/TranslateChannel';
import { SourceLanguageCode, TargetLanguageCode } from 'deepl-node';

const cacheTrChannels = new Collection<string, ITranslateChannel>();

const fetchOrCreateTrChannel = async (
  channelId: string,
  guildId: string,
  sourceLang: string,
  targetLang: string
) => {
  // fetch from db, if not exists, create new
  const channel = await TranslateChannel.findOne({ id: channelId });
  if (channel) return channel;

  const newChannel = new TranslateChannel({
    id: channelId,
    guildId,
    sourceLang,
    targetLang,
  });
  await newChannel.save();
  return newChannel;
};

const set = async (
  channelId: string,
  guildId: string,
  sourceLang: string,
  targetLang: string
) => {
  // update in db
  const channel = await fetchOrCreateTrChannel(
    channelId,
    guildId,
    sourceLang,
    targetLang
  );
  channel.overwrite({
    id: channelId,
    guildId: channel.guildId,
    sourceLang,
    targetLang,
    createdAt: channel.createdAt,
  });
  await channel.save();

  // update cache
  cacheTrChannels.set(channelId, {
    _id: channel._id,
    id: channel.id,
    guildId: channel.guildId,
    targetLang: channel.targetLang as TargetLanguageCode,
    sourceLang: channel.sourceLang as SourceLanguageCode,
    createdAt: channel.createdAt,
  });
};

const get = async (channelId: string) => {
  // get from cache
  const channelCache = cacheTrChannels.get(channelId);
  if (channelCache) return channelCache;

  // if not exists, fetch from db and set to cache
  const channel = await TranslateChannel.findOne({ id: channelId });
  if (!channel) return null;

  // update cache
  cacheTrChannels.set(channelId, {
    _id: channel._id,
    id: channel.id,
    guildId: channel.guildId,
    targetLang: channel.targetLang as TargetLanguageCode,
    sourceLang: channel.sourceLang as SourceLanguageCode,
    createdAt: channel.createdAt,
  });
  return cacheTrChannels.get(channelId)!;
};

const unset = async (channelId: string) => {
  // check if it exists first
  if ((await get(channelId)) == null) return;

  // delete from db
  await TranslateChannel.deleteOne({ id: channelId });

  // remove from cache
  cacheTrChannels.delete(channelId);
};

export default { set, unset, get };
