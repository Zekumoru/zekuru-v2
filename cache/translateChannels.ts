import { Collection } from 'discord.js';
import type { ITranslateChannel } from '../models/TranslateChannel';
import TranslateChannel from '../models/TranslateChannel';
import { SourceLanguageCode, TargetLanguageCode } from 'deepl-node';

const channels = new Collection<string, ITranslateChannel>();

const getChannel = async (
  channelId: string,
  guildId: string,
  sourceLang: string,
  targetLang: string
) => {
  // fetch from db, if not exists, create new
  const channel = await TranslateChannel.findOne({ id: channelId });
  if (channel) return channel;

  const newChannel = new TranslateChannel({
    id: channelId,
    guildId,
    sourceLang,
    targetLang,
  });
  await newChannel.save();
  return newChannel;
};

const set = async (
  channelId: string,
  guildId: string,
  sourceLang: string,
  targetLang: string
) => {
  // update in db
  const channel = await getChannel(channelId, guildId, sourceLang, targetLang);
  channel.overwrite({
    id: channelId,
    guildId: channel.guildId,
    sourceLang,
    targetLang,
    createdAt: channel.createdAt,
  });
  await channel.save();

  // update cache
  channels.set(channelId, {
    _id: channel._id,
    id: channel.id,
    guildId: channel.guildId,
    targetLang: channel.targetLang as TargetLanguageCode,
    sourceLang: channel.sourceLang as SourceLanguageCode,
    createdAt: channel.createdAt,
  });
};

const get = async (channelId: string) => {
  // get from cache
  const channelCache = channels.get(channelId);
  if (channelCache) return channelCache;

  // if not exists, fetch from db and set to cache
  const channel = await TranslateChannel.findOne({ id: channelId });
  if (!channel) return null;

  // update cache
  channels.set(channelId, {
    _id: channel._id,
    id: channel.id,
    guildId: channel.guildId,
    targetLang: channel.targetLang as TargetLanguageCode,
    sourceLang: channel.sourceLang as SourceLanguageCode,
    createdAt: channel.createdAt,
  });
  return channels.get(channelId)!;
};

const unset = async (channelId: string) => {
  // check if it exists first
  if ((await get(channelId)) == null) return;

  // delete from db
  await TranslateChannel.deleteOne({ id: channelId });

  // remove from cache
  channels.delete(channelId);
};

export default { set, unset, get };
