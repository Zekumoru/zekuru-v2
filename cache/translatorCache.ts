import { AuthorizationError, Translator, TranslatorOptions } from 'deepl-node';
import { Collection } from 'discord.js';
import GuildKey, { IGuildKey } from '../models/GuildKey';
import { loadLanguages } from '../translation/languages';
import { errorDebug } from '../utils/logger';
import { decrypt, encrypt } from '../utils/crypt';

export const translatorOptions: TranslatorOptions = {
  appInfo: {
    appName: 'Zekuru-v2 Demo',
    appVersion: '1.0.0',
  },
  minTimeout: 500, // 500 ms
  maxRetries: 10,
};

const cacheTranslators = new Collection<string, Translator>();

const fetchOrCreateGuildKey = async (guildId: string, hashedApiKey: string) => {
  // fetch from db, if not exists, create new
  const guildKey = await GuildKey.findOne({ id: guildId });
  if (guildKey) return [guildKey, false] as const;

  // save api key to db
  const newGuildKey = new GuildKey({
    id: guildId,
    key: hashedApiKey,
  });
  await newGuildKey.save();

  return [newGuildKey, true] as const;
};

const set = async (guildId: string, apiKey: string) => {
  // create translator client
  const translator = new Translator(apiKey, translatorOptions);

  // this should throw error if invalid api key
  await translator.getUsage();
  await loadLanguages(translator);

  // encrypt api key
  const encryptedApiKey = encrypt(apiKey);

  // update in db
  const [guildKey, isNewlyCreated] = await fetchOrCreateGuildKey(
    guildId,
    encryptedApiKey
  );

  if (!isNewlyCreated) {
    guildKey.overwrite({
      id: guildId,
      key: encryptedApiKey,
      createdAt: guildKey.createdAt,
    } as IGuildKey);
    await guildKey.save();
  }

  // update cache
  cacheTranslators.set(guildId, translator);
};

const get = async (guildId: string) => {
  // get from cache
  const translatorCache = cacheTranslators.get(guildId);
  if (translatorCache) return translatorCache;

  // if not exists in cache, fetch from api key from db
  // then create new translator and set to cache
  const guildKey = await GuildKey.findOne({ id: guildId });
  if (!guildKey) return;

  try {
    const apiKey = decrypt(guildKey.key);
    const translator = new Translator(apiKey, translatorOptions);

    // load languages
    await loadLanguages(translator);

    // update cache
    cacheTranslators.set(guildId, translator);
    return translator;
  } catch (error) {
    // delete invalid api key from db and "sign out" user by removing it from cache
    if (error instanceof AuthorizationError) {
      await unset(guildId);
    }
    errorDebug(error);
  }
};

const unset = async (guildId: string) => {
  // check if it exists first
  if ((await get(guildId)) == null) return;

  // delete from db
  await GuildKey.deleteOne({ id: guildId });

  // remove from cache
  cacheTranslators.delete(guildId);
};

export default { set, unset, get };
