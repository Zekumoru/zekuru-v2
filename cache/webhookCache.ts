import { BaseGuildTextChannel, Collection, Webhook } from 'discord.js';

// webhooks are created using guild id so there's
// only one webhook for every server

// key is channel id, value is webhook
const cacheWebhookIds = new Collection<string, Webhook>();

const get = async (channel: BaseGuildTextChannel) => {
  // find first if the webhook is in cache
  const webhookCache = cacheWebhookIds.get(channel.id);
  if (webhookCache) return webhookCache;

  // otherwise find the webhook and put it in cache
  const webhooks = await channel.fetchWebhooks();
  const webhook = webhooks.find(
    // as stated before, one webhook per guild
    (webhook) => webhook.guildId === channel.guildId
  );
  if (webhook) {
    cacheWebhookIds.set(channel.id, webhook);
    return webhook;
  }

  // ultimately, create a new webhook
  const newWebhook = await channel.createWebhook({ name: `Zekuru-v2 Webhook` });
  cacheWebhookIds.set(channel.id, newWebhook);
  return newWebhook;
};

export default {
  get,
};
