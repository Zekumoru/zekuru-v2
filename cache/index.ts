import channelLinkCache from './channelLinkCache';
import translateChannelCache from './translateChannelCache';
import translatorCache from './translatorCache';
import webhookCache from './webhookCache';

export default {
  channelLink: channelLinkCache,
  translateChannel: translateChannelCache,
  translator: translatorCache,
  webhook: webhookCache,
};
