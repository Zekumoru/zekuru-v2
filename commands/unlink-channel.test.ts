import unlinkChannel from './unlink-channel';
import {
  resetSampleChannelLinks,
  sampleChannelLinks,
  sampleTranslateChannels,
} from '../cache/__mocks__/sample-data';
import createMockInteraction from '../test/createMockChatInputCommandInteraction';
import cache from '../cache';

jest.mock('../cache');

describe('/unlink-channel command', () => {
  const mockCache = jest.mocked(cache);

  afterEach(() => {
    resetSampleChannelLinks();
    jest.clearAllMocks();
  });

  it(`should notify user that this command is only for servers`, async () => {
    const interaction = createMockInteraction({ guildId: undefined });

    await unlinkChannel.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user that the channel they provided isn't a translate channel`, async () => {
    const interaction = createMockInteraction({ channelId: '000' });

    await unlinkChannel.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user that the channel they provided isn't linked with any channels`, async () => {
    const interaction = createMockInteraction({ channelId: '100' });

    await unlinkChannel.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should unlink the provided translate channel from all of its links`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getChannel.mockReturnValue({ id: '100' } as any);
    sampleChannelLinks[0].links = [
      sampleTranslateChannels[1],
      sampleTranslateChannels[2],
    ] as any;
    sampleChannelLinks[1].links = [
      sampleTranslateChannels[0],
      sampleTranslateChannels[2],
    ] as any;
    sampleChannelLinks[2].links = [
      sampleTranslateChannels[0],
      sampleTranslateChannels[1],
    ] as any;

    await unlinkChannel.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    expect(mockCache.channelLink.update.mock.calls[0][0]).toMatchSnapshot();
    expect(mockCache.channelLink.update.mock.calls[1][0]).toMatchSnapshot();
    expect(mockCache.channelLink.update.mock.calls[2][0]).toMatchSnapshot();
    expect(mockCache.channelLink.delete.mock.calls[0][0]).toMatchSnapshot();
    expect(sampleChannelLinks[0].links).toHaveLength(0);
    expect(sampleChannelLinks[1].links).not.toContain(
      sampleTranslateChannels[0]
    );
    expect(sampleChannelLinks[2].links).not.toContain(
      sampleTranslateChannels[0]
    );
  });
});
