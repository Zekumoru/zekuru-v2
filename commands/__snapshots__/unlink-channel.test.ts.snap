// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/unlink-channel command should notify user that the channel they provided isn't a translate channel 1`] = `
{
  "content": "Cannot unlink! <#000> is not a translate channel!",
}
`;

exports[`/unlink-channel command should notify user that the channel they provided isn't linked with any channels 1`] = `
{
  "content": "Cannot unlink! <#100> is not linked with any channels!",
}
`;

exports[`/unlink-channel command should notify user that this command is only for servers 1`] = `
{
  "content": "This command is only available on servers.",
}
`;

exports[`/unlink-channel command should unlink the provided translate channel from all of its links 1`] = `
{
  "content": "<#100> is now unlinked!",
}
`;

exports[`/unlink-channel command should unlink the provided translate channel from all of its links 2`] = `
{
  "guildId": "guild-id",
  "id": "200",
  "links": [
    {
      "guildId": "guild-id",
      "id": "300",
      "sourceLang": "zh",
      "targetLang": "zh",
    },
  ],
}
`;

exports[`/unlink-channel command should unlink the provided translate channel from all of its links 3`] = `
{
  "guildId": "guild-id",
  "id": "100",
  "links": [],
}
`;

exports[`/unlink-channel command should unlink the provided translate channel from all of its links 4`] = `
{
  "guildId": "guild-id",
  "id": "300",
  "links": [
    {
      "guildId": "guild-id",
      "id": "200",
      "sourceLang": "ja",
      "targetLang": "ja",
    },
  ],
}
`;

exports[`/unlink-channel command should unlink the provided translate channel from all of its links 5`] = `"100"`;
